From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: crackerjacques <jack@supremeoverlordjabs.co>
Date: Wed, 29 Jul 2026 19:19:21 +0200
Subject: [ARCHEOLOGY] board: anbernic-rg-vita-pro - fix eMMC HS400 timeouts
 (rk3576 DLL calibration)

> X-Git-Archeology: > recovered message: > Add a dwcmshc patch that ports the vendor rk3576 HS400 DLL tap
> X-Git-Archeology: > recovered message: > calibration (revision 1 only): read the locked DLL value from
> X-Git-Archeology: > recovered message: > DLL_STATUS0 and feed it back as the tap-value select on every DLL clock
> X-Git-Archeology: > recovered message: > line, use taps 7/7/5 and BOTH_CLK_EDGE on CMDOUT.
> X-Git-Archeology: > recovered message: > Without it the mainline HS400 timing has almost no margin on this board.
> X-Git-Archeology: > recovered message: > Under sustained I/O (kernel compile, large apt transaction) the eMMC
> X-Git-Archeology: > recovered message: > stalls with "mmc0: Timeout waiting for hardware interrupt" on
> X-Git-Archeology: > recovered message: > CMD13/CMD18/CMD24 and grinds to a halt. Forcing HS200 makes it vanish,
> X-Git-Archeology: > recovered message: > which pins it on the HS400 timing rather than the card (pristine
> X-Git-Archeology: > recovered message: > PRE_EOL/LIFE_TIME, clean fsck). With the calibration the card stays in
> X-Git-Archeology: > recovered message: > HS400-ES at 200MHz and the timeouts are gone.
> X-Git-Archeology: > recovered message: > Signed-off-by: crackerjacques <jack@supremeoverlordjabs.co>
> X-Git-Archeology: - Revision 2cfa897ddbb543b22945951ede51d64e5808ff9e: https://github.com/armbian/build/commit/2cfa897ddbb543b22945951ede51d64e5808ff9e
> X-Git-Archeology:   Date: Wed, 29 Jul 2026 19:19:21 +0200
> X-Git-Archeology:   From: crackerjacques <jack@supremeoverlordjabs.co>
> X-Git-Archeology:   Subject: board: anbernic-rg-vita-pro - fix eMMC HS400 timeouts (rk3576 DLL calibration)
> X-Git-Archeology:
---
 drivers/mmc/host/sdhci-of-dwcmshc.c | 44 ++++++++--
 1 file changed, 38 insertions(+), 6 deletions(-)

diff --git a/drivers/mmc/host/sdhci-of-dwcmshc.c b/drivers/mmc/host/sdhci-of-dwcmshc.c
index 111111111111..222222222222 100644
--- a/drivers/mmc/host/sdhci-of-dwcmshc.c
+++ b/drivers/mmc/host/sdhci-of-dwcmshc.c
@@ -117,6 +117,13 @@
 #define DLL_CMDOUT_TAPNUM_FROM_SW	BIT(24)
 #define DLL_CMDOUT_SRC_CLK_NEG		BIT(28)
 #define DLL_CMDOUT_EN_SRC_CLK_NEG	BIT(29)
+#define DLL_CMDOUT_BOTH_CLK_EDGE	BIT(30)
+#define DLL_TAP_VALUE_SEL		BIT(25)
+#define DLL_TAP_VALUE_OFFSET		8
+/* HS400 DLL taps used by revision 1 (rk3576), taken from the vendor driver */
+#define DLL_TXCLK_TAPNUM_RK3576		0x7
+#define DLL_CMDOUT_TAPNUM_RK3576	0x7
+#define DLL_STRBIN_TAPNUM_RK3576	0x5
 
 #define DLL_LOCK_WO_TMOUT(x) \
 	((((x) & DWCMSHC_EMMC_DLL_LOCKED) == DWCMSHC_EMMC_DLL_LOCKED) && \
@@ -756,7 +763,7 @@ static void dwcmshc_rk3568_set_clock(struct sdhci_host *host, unsigned int clock
 	struct rk35xx_priv *priv = dwc_priv->priv;
 	const struct rockchip_pltfm_data *rockchip_pdata = to_pltfm_data(dwc_priv, rockchip);
 	u8 txclk_tapnum = DLL_TXCLK_TAPNUM_DEFAULT;
-	u32 extra, reg;
+	u32 extra, reg, dll_lock_value;
 	int err;
 
 	host->mmc->actual_clock = 0;
@@ -842,23 +849,40 @@ static void dwcmshc_rk3568_set_clock(struct sdhci_host *host, unsigned int clock
 		goto enable_clk;
 	}
 
+	dll_lock_value = ((sdhci_readl(host, DWCMSHC_EMMC_DLL_STATUS0) & 0xFF) * 2) & 0xFF;
+
 	extra = 0x1 << 16 | /* tune clock stop en */
 		0x3 << 17 | /* pre-change delay */
 		0x3 << 19;  /* post-change delay */
 	sdhci_writel(host, extra, dwc_priv->vendor_specific_area1 + DWCMSHC_EMMC_ATCTRL);
 
+	/*
+	 * Revision 1 (rk3576) feeds the measured DLL lock value back as the
+	 * tap-value select on every DLL clock line. Without that calibration
+	 * the HS400 timing has little margin and sustained I/O stalls with
+	 * "Timeout waiting for hardware interrupt".
+	 */
+	if (rockchip_pdata->revision == 1) {
+		extra = DWCMSHC_EMMC_DLL_DLYENA |
+			DLL_TAP_VALUE_SEL |
+			dll_lock_value << DLL_TAP_VALUE_OFFSET;
+		sdhci_writel(host, extra, DWCMSHC_EMMC_DLL_RXCLK);
+	}
+
 	if (host->mmc->ios.timing == MMC_TIMING_MMC_HS200 ||
 	    host->mmc->ios.timing == MMC_TIMING_MMC_HS400)
 		txclk_tapnum = priv->txclk_tapnum;
 
 	if (rockchip_pdata->revision == 1 && host->mmc->ios.timing == MMC_TIMING_MMC_HS400) {
-		txclk_tapnum = DLL_TXCLK_TAPNUM_90_DEGREES;
+		txclk_tapnum = DLL_TXCLK_TAPNUM_RK3576;
 
 		extra = DLL_CMDOUT_SRC_CLK_NEG |
-			DLL_CMDOUT_EN_SRC_CLK_NEG |
+			DLL_CMDOUT_BOTH_CLK_EDGE |
 			DWCMSHC_EMMC_DLL_DLYENA |
-			DLL_CMDOUT_TAPNUM_90_DEGREES |
-			DLL_CMDOUT_TAPNUM_FROM_SW;
+			DLL_CMDOUT_TAPNUM_RK3576 |
+			DLL_CMDOUT_TAPNUM_FROM_SW |
+			DLL_TAP_VALUE_SEL |
+			dll_lock_value << DLL_TAP_VALUE_OFFSET;
 		sdhci_writel(host, extra, DECMSHC_EMMC_DLL_CMDOUT);
 	}
 
@@ -866,11 +890,19 @@ static void dwcmshc_rk3568_set_clock(struct sdhci_host *host, unsigned int clock
 		DLL_TXCLK_TAPNUM_FROM_SW |
 		DLL_RXCLK_NO_INVERTER << DWCMSHC_EMMC_DLL_RXCLK_SRCSEL |
 		txclk_tapnum;
+	if (rockchip_pdata->revision == 1)
+		extra |= DLL_TAP_VALUE_SEL |
+			 dll_lock_value << DLL_TAP_VALUE_OFFSET;
 	sdhci_writel(host, extra, DWCMSHC_EMMC_DLL_TXCLK);
 
 	extra = DWCMSHC_EMMC_DLL_DLYENA |
-		DLL_STRBIN_TAPNUM_DEFAULT |
 		DLL_STRBIN_TAPNUM_FROM_SW;
+	if (rockchip_pdata->revision == 1)
+		extra |= DLL_STRBIN_TAPNUM_RK3576 |
+			 DLL_TAP_VALUE_SEL |
+			 dll_lock_value << DLL_TAP_VALUE_OFFSET;
+	else
+		extra |= DLL_STRBIN_TAPNUM_DEFAULT;
 	sdhci_writel(host, extra, DWCMSHC_EMMC_DLL_STRBIN);
 
 enable_clk:
-- 
Armbian

